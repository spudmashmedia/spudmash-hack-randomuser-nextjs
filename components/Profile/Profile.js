/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Spudmash Media Pty Ltd
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import React from "react";
import {
  Avatar,
  Button,
  Popover,
  Typography,
  Card,
  CardContent,
  CardActionArea,
  CardMedia,
  List,
  ListItem,
  ListItemIcon,
  ListItemText
} from "@material-ui/core";
import Phone from "@material-ui/icons/Phone";
import Mail from "@material-ui/icons/Mail";
import LocationCity from "@material-ui/icons/LocationCity";
import Person from "@material-ui/icons/Person";
import { makeStyles } from "@material-ui/core/styles";
import PopupState, { bindTrigger, bindPopover } from "material-ui-popup-state";
import { useSpring, animated } from "react-spring";

const useStyles = makeStyles(theme => ({
  root: {
    display: "flex",
    "& > *": {
      margin: theme.spacing(1)
    }
  },
  small: {
    width: theme.spacing(3),
    height: theme.spacing(3)
  },
  large: {
    width: theme.spacing(7),
    height: theme.spacing(7)
  }
}));

// NOTE: for react-spring to work, need to wrap the target component with "animated" hoc
const AnimatedAvatar = animated(Avatar);

let Profile = props => {
  const classes = useStyles();
  const { data, springSize } = props;

  const calc = (x, y) => [
    -(y - window.innerHeight / 2) / 20,
    (x - window.innerWidth / 2) / 20,
    springSize ? springSize : 1.1
  ];
  const trans = (x, y, s) =>
    `perspective(600px) rotateX(${x}deg) rotateY(${y}deg) scale(${s})`;

  const [springState, setSpringState] = useSpring(() => ({
    xys: [0, 0, 1],
    config: { mass: 5, tension: 350, friction: 40 }
  }));

  return (
    <PopupState variant="popover" popupId={data.id}>
      {popupState => (
        <div>
          <Button
            id={`btn-${data.id}`}
            title={`${data.id}:[${data.name}]`}
            {...bindTrigger(popupState)}
          >
            <AnimatedAvatar
              alt={data.name}
              src={data.image_thumb}
              onMouseMove={({ clientX: x, clientY: y }) =>
                setSpringState({ xys: calc(x, y) })
              }
              onMouseLeave={() => setSpringState({ xys: [0, 0, 1] })}
              style={{ transform: springState.xys.interpolate(trans) }}
            />
          </Button>
          <Popover
            {...bindPopover(popupState)}
            anchorOrigin={{
              vertical: "bottom",
              horizontal: "center"
            }}
            transformOrigin={{
              vertical: "top",
              horizontal: "center"
            }}
          >
            <Card>
              <CardActionArea>
                <CardMedia
                  component="img"
                  height="140"
                  image={data.image_lrg}
                  title={data.name}
                />
                <CardContent>
                  <List>
                    <ListItem>
                      <Typography variant="h5">{data.name}</Typography>
                    </ListItem>
                    <ListItem>
                      <ListItemIcon>
                        <LocationCity />
                      </ListItemIcon>
                      <ListItemText>
                        <Typography variant="body2">{data.location}</Typography>
                      </ListItemText>
                    </ListItem>
                    <ListItem>
                      <ListItemIcon>
                        <Phone />
                      </ListItemIcon>
                      <ListItemText>
                        <Typography variant="body2">{data.phone}</Typography>
                      </ListItemText>
                    </ListItem>
                    <ListItem>
                      <ListItemIcon>
                        <Mail />
                      </ListItemIcon>
                      <ListItemText>
                        <Typography variant="body2">{data.email}</Typography>
                      </ListItemText>
                    </ListItem>
                    <ListItem>
                      <ListItemIcon>
                        <Person />
                      </ListItemIcon>
                      <ListItemText>
                        <Typography variant="body2">
                          username: {data.username}
                        </Typography>
                        <Typography variant="body2">
                          password: {data.password}
                        </Typography>
                      </ListItemText>
                    </ListItem>
                  </List>
                </CardContent>
              </CardActionArea>
            </Card>
          </Popover>
        </div>
      )}
    </PopupState>
  );
};

export default Profile;
