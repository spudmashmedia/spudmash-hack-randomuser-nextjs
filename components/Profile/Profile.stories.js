/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Spudmash Media Pty Ltd
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import React from "react";
import Profile from "./Profile";
import { withA11y } from "@storybook/addon-a11y";
import { withKnobs, number } from "@storybook/addon-knobs";

export default {
  component: Profile,
  title: "Profile",
  decorators: [withA11y, withKnobs],
  excludeStories: /.*Data$/
};

export const actionData = {
  id: "77116352",
  name: "Marc Van Breemen",
  username: "LadiesMan217",
  password: "Witwicky123",
  email: "marc.vanbreemen@example.com",
  phone: "(724)-143-6221",
  location: "Rijsoord, Drenthe Netherlands",
  image_thumb: "https://randomuser.me/api/portraits/thumb/men/65.jpg",
  image_mid: "https://randomuser.me/api/portraits/med/men/65.jpg",
  image_lrg: "https://randomuser.me/api/portraits/men/65.jpg"
};

let setupSpringSizeKnob = () => {
  const label = "Spring Size";
  const defaultValue = 1.5;
  const options = {
    range: true,
    min: 1,
    max: 10,
    step: 0.1
  };
  const groupId = "GROUP-ID1";

  return number(label, defaultValue, options, groupId);
};

export const Default = () => {
  const springSize = setupSpringSizeKnob();
  return <Profile data={actionData} springSize={springSize} />;
};
