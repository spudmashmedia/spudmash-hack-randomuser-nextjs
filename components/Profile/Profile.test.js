/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Spudmash Media Pty Ltd
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import React from "react";
import Profile from "./Profile";
import renderer from "react-test-renderer";
import { actionData } from "./Profile.stories";

describe("Profile Component Snapshot Test", () => {
  test("renders correctly", () => {
    const tree = renderer.create(<Profile data={actionData} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
