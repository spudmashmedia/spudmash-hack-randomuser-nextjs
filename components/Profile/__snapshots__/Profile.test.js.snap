// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component Snapshot Test renders correctly 1`] = `
<div>
  <button
    aria-controls={null}
    aria-haspopup={true}
    className="MuiButtonBase-root MuiButton-root MuiButton-text"
    disabled={false}
    id="btn-77116352"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    title="77116352:[Marc Van Breemen]"
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circle"
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        style={
          Object {
            "transform": "perspective(600px) rotateX(0deg) rotateY(0deg) scale(1)",
          }
        }
      >
        <img
          alt="Marc Van Breemen"
          className="MuiAvatar-img"
          src="https://randomuser.me/api/portraits/thumb/men/65.jpg"
        />
      </div>
    </span>
  </button>
</div>
`;
